### PR TITLE
46 - Scheduled payments and standing order data for the Accounts API

### PR DESCRIPTION
**Pull request by matt-wills**
### Save scheduled payments and standing order data for the Accounts API

Issue: https://github.com/SecureBankingAcceleratorToolkit/SecureBankingAcceleratorToolkit/issues/46

### CODE REVIEW

File: securebanking-openbanking-uk-rs-obie-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_0/domesticscheduledpayments/DomesticScheduledPaymentsApi.java [hunk 1]
```diff
@@ -80,8 +80,11 @@
             @ApiParam(value = "A detached JWS signature of the body of the payload.", required = true)
             @RequestHeader(value = "x-jws-signature", required = true) String xJwsSignature,
 
+            @ApiParam(value = "The ID of the account that the payment is being made from.")
+            @RequestHeader(value = "x-ob-account-id", required = true) String xAccountId,
```
Comment (matt-wills) [inline securebanking-openbanking-uk-rs-obie-api/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_0/domesticscheduledpayments/DomesticScheduledPaymentsApi.java:84]: Potentially controversial adding this custom header for only the scheduled payment & standing order endpoints, but without this I can't link the data to the relevant account.

File: securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/frequency/FrequencyService.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+package com.forgerock.securebanking.openbanking.uk.rs.service.frequency;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeConstants;
+import org.springframework.stereotype.Service;
+
+import java.util.regex.Matcher;
+
+@Service
+public class FrequencyService {
```
Comment (matt-wills) [inline securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/frequency/FrequencyService.java:25]: Copied from `openbanking-aspsp`.

File: securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/frequency/FrequencyType.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.securebanking.openbanking.uk.rs.service.frequency;
+
+import java.util.regex.Pattern;
+
+public enum FrequencyType {
```
Comment (matt-wills) [inline securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/frequency/FrequencyType.java:20]: Copied from `openbanking-aspsp`.

File: securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/frequency/QuarterType.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.securebanking.openbanking.uk.rs.service.frequency;
+
+import org.joda.time.DateTime;
+
+public enum QuarterType {
```
Comment (matt-wills) [inline securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/frequency/QuarterType.java:20]: Copied from `openbanking-aspsp`.

File: securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/scheduledpayment/ScheduledPaymentService.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    public FRScheduledPayment createScheduledPayment(FRScheduledPaymentData scheduledPaymentData) {
+        log.debug("Create a scheduled payment in the repository: {}", scheduledPaymentData);
+
+        FRScheduledPayment frScheduledPayment = FRScheduledPayment.builder()
+                .id(UUID.randomUUID().toString())
+                .scheduledPayment(scheduledPaymentData)
+                .accountId(scheduledPaymentData.getAccountId())
+                .status(PENDING)
+                // TODO - do we need to persist the pispId?
+                //.pispId(pispId)
```
Comment (matt-wills) [inline securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/service/scheduledpayment/ScheduledPaymentService.java:56]: I'm not sure if we need this yet.

**Review by jorgesanchezperez — APPROVED** [review]

**Pull request merged**